### ISSUE TITLE
OSX: Inconsistent cmd+D behavior

### ISSUE THREAD

**Issue by jmeas** [atom/atom#10312]
**Operating system**: OSX El Capitan v10.11.2 (15C50)
**Atom**: v1.3.2

**Description of problem**:

`cmd+D` does not always work as intended. The expected behavior is that it selects the next instance of the highlighted occurrence. For instance, if `sandwiches` is selected, and `cmd+D` is entered, then the next instance of `sandwiches` should be selected. There should be no circumstance where any text aside from `sandwiches` become highlighted after typing `cmd+D`.

Sometimes – seemingly when `D` is pressed too quickly after pressing `cmd` – Atom gets confused and highlights _up until_ the next instance of the text.

**Steps to reproduce:**

**Update: [Follow the steps here](https://github.com/atom/atom/issues/10312#issuecomment-170034796)**
# 

Here are some screenshots:

**Expected behavior**

To reproduce these screenshots, press `cmd` and then `d` with at least 100ms of timing between them.

Image 1: this is after pressing `cmd+D` once:

![image](https://cloud.githubusercontent.com/assets/2322305/12161839/ac9c2d48-b4c8-11e5-891c-3ceb684ef8f0.png)

Image 2: this is after pressing `cmd+D` twice:

![image](https://cloud.githubusercontent.com/assets/2322305/12161828/932d761e-b4c8-11e5-9e89-ffa9dc4c552f.png)

**Unexpected behavior**

To reproduce these screenshots, type `cmd` and then `D` very quickly. A non-scientific guess is to try to get it down to ~5ms between them.

Image 1: this is after pressing `cmd+D` once:

![image](https://cloud.githubusercontent.com/assets/2322305/12161857/cbb63c3c-b4c8-11e5-8968-28c4dc6a7d58.png)

Image 2: this is after pressing `cmd+D` twice:

![image](https://cloud.githubusercontent.com/assets/2322305/12161860/d3600ea4-b4c8-11e5-90d9-64c93db8d820.png)

Is this a bug, or is it intention behavior where it functions differently when the keys are pressed too closely together? I've used Sublime Text for years, and I use this hotkey so frequently that I type it very, very quickly; ST behaves the same way whether there's an appreciable delay or not.

**Comment by lee-dohm** [comment]
I was unable to reproduce this error using Atom v1.5.0-dev-de04395 on Mac OS X 10.11.2. I'm on a brand new computer though with 16GB of RAM and an SSD, so perhaps that has something to do with it?

Also, could you show the output of the keybinding resolver when reproducing this issue? The Keybinding Resolver is <kbd>Cmd+.</kbd> on OS X and <kbd>Ctrl+.</kbd> on other platforms. It will show you what key Atom saw you press and the command that was triggered because of it.

**Comment by jmeas** [comment]
This gif is from Atom 1.4.0-beta3:

![atom](https://cloud.githubusercontent.com/assets/2322305/12200320/478ae686-b5ec-11e5-8a5a-9dc11cb019ec.gif)

**Comment by lee-dohm** [comment]
Any information on your computer? Spinning disk? Amount of RAM? Processor?

**Comment by jmeas** [comment]
![image](https://cloud.githubusercontent.com/assets/2322305/12201565/1ad402ce-b5f3-11e5-8431-684d63af6c47.png)

512gb SSD

**Comment by lee-dohm** [comment]
I was able to reproduce the issue! :metal:

In your GIF, it was apparent that you're selecting the word with the mouse/trackpad ... so here's what I did:

1. Use the trackpad to select from the end to the beginning of line 1
1. ***Without*** releasing the button, press <kbd>Cmd+D</kbd>
1. Release the button
1. Press <kbd>Cmd+D</kbd> again

What I suspect is happening is that because the mouse button is still depressed, as the cursor moves from the beginning of line 1 to the end of the next occurence, the space between is selected.

**Comment by jmeas** [comment]
Nice sleuthing, @lee-dohm ! I'll update my original post with a link to that comment, so that future reviewers can more easily find the correct way to reproduce the problem.

**Comment by lee-dohm** [comment]
I'm glad you supplied the GIF. I use the Dvorak keyboard layout which puts the <kbd>D</kbd> on the other side of the keyboard. Since I'm right-handed, I wouldn't normally be using the mouse/trackpad simultaneously with activating the key binding. :grinning:

**Comment by kepta** [comment]
Hi, I want to work on this bug.
Can any one give me some pointers as to where keyboard input os processed in source code?

**Comment by benogle** [comment]
> Can any one give me some pointers as to where keyboard input os processed in source code?

Hey, cmd-d behavior is handled [here](https://github.com/atom/find-and-replace/blob/fa4af2db1ce3612c81a55d25821424899b141c18/lib/find.coffee#L132-L140)

**Comment by kepta** [comment]
After investigating the bug,
as confirmed by @lee-dohm, I was also able to reproduce the bug.
It can be reproduced by dragging from any side but you have to do it with a second or so.
I believe it also depends on the previous selection buffer, because sometimes it just works.
@benogle  Can you please guide me on how to tackle this bug?

**Comment by kepta** [comment]
I can confirm that the problem is not with the find-and-replace package,
The problem is with the debouncing part of the code. When the bug is invoked (tapping cmd-d, quickly after selecting text with mouse),
The debounce function in underscore library is not called, but in case we avoid bug (tapping cmd-d, after a second), the debounce function is called.
And I also noticed an array of documentPollers (16) when the bug is not there, and just a single element in this array when we see the bug.

**Comment by lee-dohm** [comment]
I doubt this bug has anything to do with the debouncing logic. Here's some modified repro steps:

1. Use the trackpad to select from the end to the beginning of line 1 and don't release the button
1. Wait 30 seconds
1. ***Without*** releasing the button, press <kbd>Cmd+D</kbd>
1. Wait 30 seconds
1. Release the button
1. Wait another 30 seconds
1. Press <kbd>Cmd+D</kbd> again

It's purely an order of operations issue.

**Comment by kepta** [comment]
@lee-dohm I understand what you are saying, but its just that debounce showed up whenever there was no bug. Would it be related to mouseUp, were you able to reproduce this bug with a mouse?

**Comment by lee-dohm** [comment]
I didn't use a mouse, but I'm completely confident that it will reproduce with one.

**Comment by jmeas** [comment]
> but I'm completely confident that it will reproduce with one.

You are correct; just tested.

**Comment by kepta** [comment]
I might be wrong, 
but @jmeas if you release the mouse button and then click cmd - d, are you getting the bug?
Because I am getting it only if I don't release the left mouse button.

**Comment by jmeas** [comment]
Oh, interesting @kepta! I can sometimes reproduce a bug, which is what I think you're describing, using these steps:

1. Use the trackpad to select from the end to the beginning of line 1 and release the button
2. Make sure you released the button!
3. Wait 30 seconds
3. Press <kbd>Cmd+D</kbd>
7. Observe weird highlighted behavior

Note that this one is much harder to reproduce (seems like < 10% occurrence rate), so I'm not exactly sure what I'm doing differently between each trial. It seems like the faster I swipe with my fingers when highlighting, the more likely I am to reproduce the bug.

Is that what you mean?

With that said, I can **also** consistently reproduce the steps described here: https://github.com/atom/atom/issues/10312#issuecomment-171347042

**Comment by kepta** [comment]
@jmeas thats quite weird. Here is what I did,
I tried multiple times following the steps you just mentioned.
I selected from end to beginning of line 1, and released my trackpad.(Leaving the selection stay for 10 seconds)
and then tried Cmd+D.

But this didn't work in my case even after multiple repetitions.
I was planning to put up a theory that the bug will never happen whenever system registers 
```
onMouseUp && dragging
```
and for the case of Macbook trackpads, apple has a 200ms - 500ms  hardware delay to notify mouseUp,
that is why it is easy for me to select with trackpad and after 200-300ms press cmd -d, and get the bug.

But with the mouse, the moment onMouseUp is called in the console, I just couldn't reproduce the bug.
It seems there is a very small delay with a mouse.
Let me know what you think.
